File: dotnet/v3.2/Reports/FindAndDownloadFile.cs
/*
 * Copyright 2018 Google Inc
 *
 * Licensed under the Apache License, Version 2.0(the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Linq;
using Google.Apis.Dfareporting.v3_2;
using Google.Apis.Dfareporting.v3_2.Data;

namespace DfaReporting.Samples {
  /// <summary>
  /// This example provides an end-to-end example of how to find and download a
  /// report file.
  /// </summary>
  class FindAndDownloadFile: SampleBase {
    /// <summary>
    /// Returns a description about the code example.
    /// </summary>
    public override string Description {
      get {
        return "This example provides an end-to-end example of how to find " +
                "and download a report file.\n";
      }
    }

    /// <summary>
    /// Main method, to run this code example as a standalone application.
    /// </summary>
    /// <param name="args">The command line arguments.</param>
    public static void Main(string[] args) {
      SampleBase codeExample = new FindAndDownloadFile();
      Console.WriteLine(codeExample.Description);
      codeExample.Run(DfaReportingFactory.getInstance());
    }

    /// <summary>
    /// Run the code example.
    /// </summary>
    /// <param name="service">An initialized Dfa Reporting service object
    /// </param>
    public override void Run(DfareportingService service) {
      long profileId = long.Parse(_T("INSERT_USER_PROFILE_ID_HERE"));
      long reportId = long.Parse(_T("INSERT_REPORT_ID_HERE"));

      // 1. Find a file to download.
      File file = FindFile(service, profileId, reportId);

      if (file != null) {
        // 2. (optional) Generate browser URL.
        GenerateBrowserUrl(service, reportId, file.Id.Value);

        // 3. Directly download the file
        DirectDownloadFile(service, reportId, file.Id.Value);
      } else {
        Console.WriteLine(
            "No file found for profile ID {0} and report ID {1}.",
            profileId, reportId);
      }
    }

    private File FindFile(DfareportingService service, long profileId,
        long reportId) {
      File target = null;
      FileList files;
      String nextPageToken = null;

      do {
        // Create and execute the files list request.
        ReportsResource.FilesResource.ListRequest request =
            service.Reports.Files.List(profileId, reportId);
        request.PageToken = nextPageToken;
        files = request.Execute();

        foreach (File file in files.Items) {
          if (IsTargetFile(file)) {
            target = file;
            break;
          }
        }

        // Update the next page token.
        nextPageToken = files.NextPageToken;
      } while (target == null
          && files.Items.Any()
          && !String.IsNullOrEmpty(nextPageToken));

      if (target != null) {
        Console.WriteLine("Found file {0} with filename \"{1}\".",
            target.Id, target.FileName);
        return target;
      }

      Console.WriteLine(
          "Unable to find file for profile ID {0} and report ID {1}.",
          profileId, reportId);
      return null;
    }

    private bool IsTargetFile(File file) {
      // Provide custom validation logic here.
      // For example purposes, any file with REPORT_AVAILABLE status is
      // considered valid.
      return "REPORT_AVAILABLE".Equals(file.Status);
    }

    private void GenerateBrowserUrl(DfareportingService service, long reportId,
        long fileId) {
      File file = service.Files.Get(reportId, fileId).Execute();
      String browserUrl = file.Urls.BrowserUrl;

      Console.WriteLine("File {0} has browser URL: {1}.", file.Id, browserUrl);
    }

    private void DirectDownloadFile(DfareportingService service, long reportId,
        long fileId) {
      // Retrieve the file metadata.
      File file = service.Files.Get(reportId, fileId).Execute();

      if ("REPORT_AVAILABLE".Equals(file.Status)) {
        // Create a get request.
        FilesResource.GetRequest getRequest = service.Files.Get(reportId, fileId);

        // Optional: adjust the chunk size used when downloading the file.
        // getRequest.MediaDownloader.ChunkSize = MediaDownloader.MaximumChunkSize;

        // Execute the get request and download the file.
        using (System.IO.FileStream outFile = new System.IO.FileStream(GenerateFileName(file),
            System.IO.FileMode.Create, System.IO.FileAccess.Write)) {
          getRequest.Download(outFile);
          Console.WriteLine("File {0} downloaded to {1}", file.Id, outFile.Name);
        }
      }
    }

    private string GenerateFileName(File file) {
      // If no filename is specified, use the file ID instead.
      string fileName = file.FileName;
      if (String.IsNullOrEmpty(fileName)) {
        fileName = file.Id.ToString();
      }

      String extension = "CSV".Equals(file.Format) ? ".csv" : ".xml";

      return fileName + extension;
    }
  }
}

File: ruby/v3.2/reports/find_and_download_file.rb
#!/usr/bin/env ruby

#
# Copyright:: Copyright 2018, Google Inc. All Rights Reserved.
#
# License:: Licensed under the Apache License, Version 2.0 (the "License");
#           you may not use this file except in compliance with the License.
#           You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#           Unless required by applicable law or agreed to in writing, software
#           distributed under the License is distributed on an "AS IS" BASIS,
#           WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#           implied.
#           See the License for the specific language governing permissions and
#           limitations under the License.
#
# An end-to-end example of how to find and download a report file.

require_relative '../dfareporting_utils'

def find_and_download_file(profile_id, report_id)
  # Authenticate and initialize API service.
  service = DfareportingUtils.get_service

  # 1. Find a file to download.
  report_file = find_file(service, profile_id, report_id)

  if report_file
    # 2. (optional) Generate a browser URL.
    generate_browser_url(service, report_id, report_file.id)

    # 3. Directly download the file.
    direct_download_file(service, report_id, report_file.id)
  else
    puts format('No file found for profile ID %d and report ID %d.', profile_id, report_id)
  end
end

def find_file(service, profile_id, report_id)
  page_token = nil
  target = nil

  begin
    result = service.list_report_files(profile_id, report_id,
      page_token: page_token)

    result.items.each do |file|
      if is_target_file(file)
        target = file
        break
      end
    end

    page_token = (result.next_page_token if target.nil? && result.items.any?)
  end until page_token.to_s.empty?

  if target
    puts format('Found file %s with filename "%s".', target.id, target.file_name)
    return target
  end

  puts format('Unable to find file for profile ID %d and report ID %d.', profile_id, report_id)
  nil
end

def is_target_file(file)
  # Provide custom validation logic here.
  # For example purposes, any available file is considered valid.
  file.status == 'REPORT_AVAILABLE'
end

def generate_browser_url(service, report_id, file_id)
  report_file = service.get_file(report_id, file_id)
  browser_url = report_file.urls.browser_url

  puts format('File %s has browser URL: %s.', report_file.id, browser_url)
end

def direct_download_file(service, report_id, file_id)
  # Retrieve the file metadata.
  report_file = service.get_file(report_id, file_id)

  if report_file.status == 'REPORT_AVAILABLE'
    # Prepare a local file to download the report contents to.
    File.open(generate_file_name(report_file), 'w') do |out_file|
      # Execute the download request. Providing a download destination
      # retrieves the file contents rather than the file metadata.
      service.get_file(report_id, file_id, download_dest: out_file)

      puts format('File %s downloaded to %s', file_id, File.absolute_path(out_file.path))
    end
  end
end

def generate_file_name(report_file)
  file_name = report_file.file_name
  # If no filename is specified, use the file ID instead.
  file_name = report_file.id.to_s if file_name.empty?
  extension = report_file.format == 'CSV' ? '.csv' : '.xml'
  file_name + extension
end

if $PROGRAM_NAME == __FILE__
  # Retrieve command line arguments.
  args = DfareportingUtils.get_arguments(ARGV, :profile_id, :report_id)

  find_and_download_file(args[:profile_id], args[:report_id])
end
